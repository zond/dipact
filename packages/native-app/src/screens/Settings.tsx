import React from "react";
import { StyleProp, StyleSheet, Text, View } from "react-native";
import { Button } from "react-native-elements";
import { useTheme } from "../hooks";
import Switch from "../components/Switch";
import TextInput from "../components/TextInput";

const values = {
	enableEmailNotifications: false,
	enablePushNotifications: true,
	phaseDeadline: "60",
};

const onValueChange = () => {};

const useStyles = (): StyleProp<any> => {
	const theme = useTheme();
	return {
		formContainer: {
			margin: theme.spacing(1),
		},
		formLabel: {
			color: theme.palette.text.light,
		},
		formSection: {
			marginTop: theme.spacing(1),
			marginBottom: theme.spacing(1),
		},
		buttonContainer: {
			display: "flex",
			flexDirection: "row",
			justifyContent: "space-between",
			margin: theme.spacing(1),
		},
		buttonText: {
			color: theme.palette.secondary.main,
		},
	};
};

const Settings = () => {
	const styles = useStyles();

	return (
		<View style={styles.formContainer}>
			<View style={styles.formSection}>
				<Text style={styles.formLabel}>Notifications</Text>
				<Switch
					value={values.enablePushNotifications}
					label="Push notifications"
					onValueChange={onValueChange}
				/>
				<Switch
					value={values.enableEmailNotifications}
					label="Email notifications"
					onValueChange={onValueChange}
				/>
			</View>
			<TextInput
				label="Phase deadline reminder"
				value={values.phaseDeadline}
				helpText="In minutes. 0 = off"
				keyboardType="numeric"
			/>
			<View style={styles.buttonContainer}>
				<Button title="Reset settings" onPress={() => {}} type="clear" />
				<Button
					title="Save changes"
					onPress={() => {}}
					titleStyle={styles.buttonText}
				/>
			</View>
		</View>
	);
};

export default Settings;
