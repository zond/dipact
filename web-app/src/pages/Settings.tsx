/* eslint-disable no-restricted-globals */
import "firebase/messaging";
import React, { useState } from "react";
import {
	Button,
	Container,
	FormControl,
	FormControlLabel,
	InputLabel,
	MenuItem,
	Select,
	Switch,
	TextField,
	Typography,
	makeStyles,
	IconButton,
	Tooltip,
} from "@material-ui/core";
import useRegisterPageView from "../hooks/useRegisterPageview";
import ResetSettingsDialog from "../components/ResetSettingsDialog";
import useSettingsForm from "../hooks/useSettingsForm";
import NavBar from "../components/NavBar";
import Loading from "../components/Loading";
import { UndoIcon } from "../icons";
import ErrorPage from "../components/ErrorPage";

const useStyles = makeStyles((theme) => {
	return {
		root: {
			"& section": {
				paddingTop: theme.spacing(2),
			},
		},
		textField: {
			maxWidth: theme.spacing(24),
		},
		buttons: {
			display: "flex",
			justifyContent: "space-between",
		},
		variantSelect: {
			display: "flex",
			flexDirection: "column",
			gap: theme.spacing(1),
		},
		nations: {
			display: "flex",
			justifyContent: "space-between",
			alignItems: "center",
			"& > div": {
				display: "flex",
				alignItems: "center",
			},
		},
	};
});

interface SettingsProps {
	useSettingsFormDI?: typeof useSettingsForm;
}

const Settings = ({
	useSettingsFormDI = useSettingsForm,
}: SettingsProps): React.ReactElement => {
	const classes = useStyles();
	useRegisterPageView("SettingsDialog"); // NOTE rename?
	const userId = "123456789";
	const {
		edited,
		editedColors,
		resetColor,
		values,
		handleChange,
		handleSubmit,
		handleResetSettings,
		isLoading,
		isError,
		variant,
		handleChangeVariant,
		disabledFields,
		helpText,
		loaded,
	} = useSettingsFormDI(userId);
	const [resetSettingsDialogOpen, setResetSettingsDialogOpen] = useState(false);

	const onClickResetSettings = () => setResetSettingsDialogOpen(true);
	return (
		<>
			<NavBar title="Settings" />
			{isError ? (
				<ErrorPage />
			) : !loaded ? (
				<Loading />
			) : (
				<Container className={classes.root}>
					<form onSubmit={handleSubmit}>
						<section>
							<Typography variant="subtitle2">Notifications</Typography>
							<div>
								<FormControlLabel
									label="Push notifications"
									control={
										<Switch
											name="enablePushNotifications"
											checked={values.enablePushNotifications}
											disabled={disabledFields.enablePushNotifications}
											onChange={handleChange}
										/>
									}
								/>
							</div>
							<div>
								<FormControlLabel
									label="Email notifications"
									control={
										<Switch
											name="enableEmailNotifications"
											checked={values.enableEmailNotifications}
											disabled={disabledFields.enableEmailNotifications}
											onChange={handleChange}
										/>
									}
								/>
							</div>
						</section>
						<div>
							<TextField
								label="Phase deadline reminder"
								name="phaseDeadline"
								type="number"
								inputProps={{ min: 0 }}
								fullWidth
								className={classes.textField}
								disabled={disabledFields.phaseDeadline}
								helperText={helpText.phaseDeadline}
								margin="dense"
								value={values.phaseDeadline}
								onChange={handleChange}
							/>
						</div>

						<section>
							<Typography variant="subtitle2">Custom nation colours</Typography>
							<FormControl>
								<InputLabel shrink id="variantinputlabel">
									Variant
								</InputLabel>
								<Select
									fullWidth
									name="variant"
									labelId="variantinputlabel"
									value={variant}
									onChange={handleChangeVariant}
								>
									{Object.keys(values.colors).map((variantName) => (
										<MenuItem key={variantName} value={variantName}>
											{variantName}
										</MenuItem>
									))}
								</Select>
							</FormControl>
							<div className={classes.variantSelect}>
								{values.colors[variant] &&
									Object.entries(values.colors[variant]).map(
										([nation, color]) => (
											<div className={classes.nations} key={nation}>
												<Typography>{nation}</Typography>
												<div>
													{editedColors[variant][nation] && (
														<Tooltip title="Undo">
															<IconButton
																onClick={() => resetColor(variant, nation)}
															>
																<UndoIcon />
															</IconButton>
														</Tooltip>
													)}
													<input
														type="color"
														name={`colors.${variant}.${nation}`}
														disabled={disabledFields.colors}
														value={color}
														onChange={handleChange}
													/>
												</div>
											</div>
										)
									)}
							</div>
						</section>
						<section>
							<div className={classes.buttons}>
								<Button
									variant="text"
									onClick={onClickResetSettings}
									disabled={isLoading}
								>
									Reset settings
								</Button>
								<Button
									type="submit"
									variant="contained"
									color="primary"
									disabled={isLoading || !edited}
								>
									Save Changes
								</Button>
							</div>
						</section>
						<ResetSettingsDialog
							open={resetSettingsDialogOpen}
							close={() => setResetSettingsDialogOpen(false)}
							handleResetSettings={handleResetSettings}
							isLoading={isLoading}
						/>
					</form>
				</Container>
			)}
		</>
	);
};

export default Settings;
