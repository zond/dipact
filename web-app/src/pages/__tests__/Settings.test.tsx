import React from "react";
import ReactDOM from "react-dom";
import Enzyme from "enzyme";
import Settings from "../Settings";
import { Scenarios, createSettingsFormDI } from "../Settings.scenarios";
import Adapter from "@wojtekmaj/enzyme-adapter-react-17";
import { MemoryRouter } from "react-router-dom";

Enzyme.configure({ adapter: new Adapter() });

interface WrappedSettingsProps {
	scenario: typeof Scenarios.Loaded;
}

const WrappedSettings = ({ scenario }: WrappedSettingsProps) => {
	const useDI = createSettingsFormDI(scenario);
	return (
		<MemoryRouter>
			<Settings useSettingsFormDI={useDI} />
		</MemoryRouter>
	);
};

describe("Snapshot tests", () => {
	let container: HTMLDivElement;
	beforeEach(() => {
		// setup a DOM element as a render target
		container = document.createElement("div");
		document.body.appendChild(container);
	});
	test("Not Loaded", () => {
		const rendered = ReactDOM.render(
			<WrappedSettings scenario={Scenarios.NotLoaded} />,
			container
		);
		expect(rendered).toMatchSnapshot();
	});
	test("IsLoading", () => {
		const rendered = ReactDOM.render(
			<WrappedSettings scenario={Scenarios.IsLoading} />,
			container
		);
		expect(rendered).toMatchSnapshot();
	});
	test("Loaded", () => {
		const rendered = ReactDOM.render(
			<WrappedSettings scenario={Scenarios.Loaded} />,
			container
		);
		expect(rendered).toMatchSnapshot();
	});
	test("TokenEnabled", () => {
		const rendered = ReactDOM.render(
			<WrappedSettings scenario={Scenarios.TokenEnabled} />,
			container
		);
		expect(rendered).toMatchSnapshot();
	});
	test("Unedited", () => {
		const rendered = ReactDOM.render(
			<WrappedSettings scenario={Scenarios.Unedited} />,
			container
		);
		expect(rendered).toMatchSnapshot();
	});
});

// describe("Settings", () => {
// 	test("Create component", () => {
// 		const div = document.createElement("div");
// 		ReactDOM.render(<WrappedSettings />, div);
// 	});

// 	test("Mount component", () => {
// 		mount(<WrappedSettings />);
// 	});

// 	test("Colors for classical variant are visible", () => {});

// 	test("Home button", () => {});

// 	test("Changing variant changes nations", () => {});

// 	test("Toggle push notifications disabled", () => {});

// 	test("Toggle push notifications", () => {});

// 	test("Toggle email notifications disabled", () => {});

// 	test("Toggle email notifications", () => {});

// 	test("Reset color", () => {});

// 	test("Color edited", () => {});

// 	test("Submit form", () => {});

// 	test("Success response closes dialog", () => {});

// 	test("Error response closes dialog", () => {});

// 	test("Success response shows message", () => {});

// 	test("Error response shows message", () => {});

// 	test("Save changes button disabled if no changes", () => {});

// 	test("Submit form push notifications enabled", () => {});

// 	test("Submit form email notifications enabled", () => {});

// 	test("Submit form colors changed", () => {});

// 	test("Reset settings opens dialog", () => {});

// 	test("Loading", () => {});
// });
