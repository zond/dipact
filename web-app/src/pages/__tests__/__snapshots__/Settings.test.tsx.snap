// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot tests IsLoading 1`] = `null`;

exports[`Snapshot tests Loaded 1`] = `null`;

exports[`Snapshot tests Not Loaded 1`] = `null`;

exports[`Snapshot tests TokenEnabled 1`] = `null`;

exports[`Snapshot tests Unedited 1`] = `null`;
